### PR TITLE
Trace metadata during task creation

### PR DESCRIPTION
**Pull request by duanmeng**
Create a directory named `$QueryTraceBaseDir/$taskId` when a task is initiated,
if query tracing is enabled. This directory will store metadata related to the task,
including the query plan node tree, query configurations, and connector properties.

Part of #9668

### CODE REVIEW

File: velox/common/memory/Memory.cpp
Comment (xiaoxmeng) [inline velox/common/memory/Memory.cpp:417]: Can you add unit test for this? thanks!

File: velox/exec/Task.h
Comment (xiaoxmeng) [inline velox/exec/Task.h:1190]:
```
s/Dump/Records/
```

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2846]: Can you add a check ahead that traceDirectory_ is empty? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2848]: Drop . at the end? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2852]: ditto

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2851]: I think we just throw here?

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2862]:
```
std::move(nodeSet)?
```

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2860]: Shall we log the traced plan node ids here?

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2868]: Let's remove it if it exist? It could be left over from previous run? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2888]: No need to check here?

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2896]: Shall we use a local object as we don't need it after write the metadata here?

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng LGTM % minors.

File: velox/exec/Task.cpp
Comment (duanmeng) [reply velox/exec/Task.cpp:2851]: `queryTraceNodes` would be empty if we only want to record the query metadata.

File: velox/exec/Task.cpp
Comment (duanmeng) [reply velox/exec/Task.cpp:2846]: removed `traceDirectory_`.

File: velox/common/memory/Memory.cpp
Comment (duanmeng) [reply velox/common/memory/Memory.cpp:417]: added
```diff
   ASSERT_THAT(
       detailedManagerStr,
       testing::HasSubstr("__sys_spilling__ usage 0B reserved 0B peak 0B\n"));
+  ASSERT_THAT(
+      detailedManagerStr,
+      testing::HasSubstr("__sys_tracing__ usage 0B reserved 0B peak 0B\n"));
```

**Review by xiaoxmeng — COMMENTED** [review]
Can you add PR description? thanks!

File: velox/core/QueryConfig.h
Comment (xiaoxmeng) [inline velox/core/QueryConfig.h:350]: Can you update config.rst? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:2850]: I see. Then we don't need this log? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2852]: Maybe have a constructor for query trace config with metadata only? Thanks!

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2865]: void? Why return the same input? This can be anonymous function or put in QueryTraceUtil.h under trace? Thanks!

File: velox/exec/trace/test/QueryTraceTest.cpp
Comment (xiaoxmeng) [inline velox/exec/trace/test/QueryTraceTest.cpp:206]: Do you have a case with no plan nodes but metadata only? Also can you add a failure test with invalid config? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng thanks for the update!

File: velox/core/QueryConfig.h
Comment (duanmeng) [reply velox/core/QueryConfig.h:350]: Got it, done.

File: velox/exec/trace/test/QueryTraceTest.cpp
Comment (duanmeng) [reply velox/exec/trace/test/QueryTraceTest.cpp:206]: >can you add a failure test with invalid config?

Sure, added.

>Do you have a case with no plan nodes but metadata only?

It would throw in `AssertQueryBuilder` if `planNode` is null (no plan nodes). Is this a valid state that a task is creation without any plan nodes?

File: velox/docs/configs.rst
Comment (xiaoxmeng) [inline velox/docs/configs.rst:708]:
```
The root directory to store the tracing data and metadata for a query
```

File: velox/docs/configs.rst
Comment (xiaoxmeng) [inline velox/docs/configs.rst:712]:
```
be trace. If it is empty, then we only trace the query metadata which includes the query plan and configs etc.
```

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:2847]:
```
s/VELOX_CHECK/VELOX_USER_CHECK/
```

File: velox/exec/Task.cpp
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:2853]: nit: can we have a ctor which only take the trace directory? Thanks!

File: velox/exec/trace/QueryTraceConfig.cpp
Comment (xiaoxmeng) [inline velox/exec/trace/QueryTraceConfig.cpp:29]: Why we need this?

File: velox/exec/trace/QueryTraceUtil.h
Comment (xiaoxmeng) [inline velox/exec/trace/QueryTraceUtil.h:24]:
```
/// Creates ...
```

File: velox/exec/trace/QueryTraceUtil.h
Comment (xiaoxmeng) [inline velox/exec/trace/QueryTraceUtil.h:25]: Can we have a unit test for this?

File: velox/exec/trace/test/QueryTraceTest.cpp
Comment (xiaoxmeng) [reply velox/exec/trace/test/QueryTraceTest.cpp:206]: Yes, let's expect an throw there? VELOX_ASSERT_USER_THROW?

File: velox/exec/trace/test/QueryTraceTest.cpp
Comment (xiaoxmeng) [inline velox/exec/trace/test/QueryTraceTest.cpp:285]:
```
s/fail/error/
```

File: velox/exec/trace/QueryTraceConfig.cpp
Comment (duanmeng) [reply velox/exec/trace/QueryTraceConfig.cpp:29]: Fixed

File: velox/exec/Task.cpp
Comment (duanmeng) [reply velox/exec/Task.cpp:2853]: fixed

File: velox/exec/trace/QueryTraceUtil.h
Comment (duanmeng) [reply velox/exec/trace/QueryTraceUtil.h:24]: Added

**Review by xiaoxmeng — APPROVED** [review]
@duanmeng LGTM. Thanks!

**Review by Yuhta — APPROVED** [review]

File: velox/exec/trace/test/CMakeLists.txt
Comment (assignUser) [inline velox/exec/trace/test/CMakeLists.txt:38]: Are you sure you need to explicitly link against all of these? I don't see any explicit `#includes` for any headers from these targets. The other targets will propagate their usage requirements, there is no need to add them manually.

File: velox/exec/trace/CMakeLists.txt
Comment (assignUser) [inline velox/exec/trace/CMakeLists.txt:15]: I see the issue for the linking error. It's an issue with the monolithic library build and it's helper functions that now need to be used. Sorry we haven't communicated that well yet. #9948 took a while to merge so the targets here were added after CI was run but before it got merged, sorry for the inconvenience.

For targets that are not standalone executables like tests or fuzzers etc. you have to use `velox_add_library` instead of `add_library` and `velox_link_libraries` instead of `target_link_libraries` now.

Comment (duanmeng) [reply velox/exec/trace/CMakeLists.txt:15]: >you have to use velox_add_library instead of add_library and velox_link_libraries instead of target_link_libraries now.

@assignUser Do you mean that I don't need to link velox_query_trace_exec explicitly in `velox/benchmarks`, `velox/benchmarks/basic`, and `velox/dwio/dwrf/test` if I use `velox_add_library(velox_query_trace_exec...` and `velox_link_libraries(velox_query_trace_exec...` in `velox/exec/trace/CMakeList.txt`.

Comment (assignUser) [reply velox/exec/trace/CMakeLists.txt:15]: @xiaoxmeng  please don't merge before this is fixed

Comment (assignUser) [reply velox/exec/trace/CMakeLists.txt]: No, you still have to link the target where you are using it. When `VELOX_MONO_LIBRARY=OFF` everything behaves as if you were using the original functions so you still need those links.  (especially for the tests and benchmark as those are unaffected by the mono build)

Comment (duanmeng) [reply velox/exec/trace/CMakeLists.txt]: >For targets that are not standalone executables like tests or fuzzers etc. you have to use velox_add_library instead of add_library and velox_link_libraries instead of target_link_libraries now.

@assignUser Thanks for your suggestion, no more linking errors and ci passed.

**Pull request closed without merging**
